### ISSUE TITLE
Write/Read zip file from/to memory buffers

### ISSUE THREAD

**Issue by mzf-guest** [Zipios/Zipios#37]
With the current 2.x version of Zipios I can't find a way to write file content to a zip file from a memory buffer and also a way to read zip file from a memory buffer (using std::ostream/std::istream for example).

In the previous version 0.1.x of Zipios, the ZipInputStreambuf/ZipOutputStreambuf/ZipCDirEntry were appropriate but they are not part of the public API anymore.
The only workaround I've found is to write buffer to a temporary file before writing them back to the ZipFile via a DirectoryCollection. Am I missing something?

[Context: I'm updating zipios reverse dependencies in Debian archive and the enigma package does some [zip file edition from/to memory buffer](https://github.com/Enigma-Game/Enigma/blob/master/src/file_zip.cc#L62) that is hard to update with zipios 2.x]

**Comment by AlexisWilke** [comment]
From what I can tell, this means all of these need to be public:

    deflateoutputstreambuf.hpp
    filterinputstreambuf.hpp
    filteroutputstreambuf.hpp
    inflateinputstreambuf.hpp
    zipcentraldirectoryentry.hpp
    zipinputstreambuf.hpp
    ziplocalentry.hpp
    zipoutputstreambuf.hpp

Will that be enough in your case, or would other additional files be required?

As a result you could use:

    DeflateOutputStreambuf(std::streambuf *outbuf)
    InflateInputStreambuf(std::streambuf *inbuf, offset_t start_pos)

Is that what you're looking at? You'd of course have to implement your own version of `streambuf` in order to be able to read from your existing memory buffer. At the same time, it would be great if it were to also work on existing streams.

**Comment by mzf-guest** [comment]
Yes one solution is to move the missing headers to the public folder.
But I like the idea to keep these classes private because they are only internal implementation.

So, one other solution is to modify the ZipFile class to be able to manipulate stream buffers directly:

- add a ZipFile constructor that takes the file content as `std::istream` (or `std::streambuf`) instead of the filename,
- add a ZipFile method to write the zip output to a `std::ostream` (like `ZipFile::write(std::ostream& outputStream)`)
- add a ZipFile method to add entry from a stream like `ZipFile::addEntry(const std::istream& fileContent, ...)` (could also be a dedicated FileEntry that could be built from a `std::istream` for its content).

Is it easy to implement? otherwise the first solution is OK.
Thanks,

**Comment by mzf-guest** [comment]
Hello!

Any news on this issue?
This is still blocking the migration to 2.x in Debian.

Best,
François

**Comment by AlexisWilke** [comment]
Hi François,

Good news, I did add some functions and pushed the results. It compiles and the old tests still pass. I am not 100% sure the new code will work, though. I did not have the time to write not test to make sure.

* _add a `ZipFile` constructor that takes the file content as `std::istream` (or `std::streambuf`) instead of the filename,_

  I did that. The `ZipFile` class now has a new constructor that takes an `std::istream`. It reuses the same code as the other constructor so I'm pretty confident it will work.

  **One Important Constraint:** The istream must be fully seekable.

* _add a `ZipFile` method to write the zip output to a `std::ostream` (like `ZipFile::write(std::ostream& outputStream)`)_

  Isn't the `static void saveCollectionToArchive(std::ostream & os, FileCollection & collection, std::string const & zip_comment = std::string());` sufficient for that purpose? The `ZipFile` is a `FileCollection` so it can be passed to that function.

* _add a ZipFile method to add entry from a stream like ZipFile::addEntry(const std::istream& fileContent, ...) (could also be a dedicated FileEntry that could be built from a std::istream for its content)._

  I added a new object I called `StreamEntry`. Create one of those and then do an `AddEntry()` on your `ZipFile` object (again, since we derive from a `FileCollection`, we have access to the `AddEntry()` function defined there).

                                    StreamEntry(
                                              std::istream & is
                                            , FilePath const & filename
                                            , std::string const & comment = std::string());

Hope this helps.

**Comment by mzf-guest** [comment]
Awesome! I'll look at it and test shortly. Thanks!

**Comment by RokerHRO** [comment]
I though I could just use the classes `ZipInputStream` and `ZipOutputStream`, or am I wrong?

**Comment by AlexisWilke** [comment]
@RokerHRO Yes. You can just use those. It's expected to work like a charm.

**Comment by mzf-guest** [comment]
> I though I could just use the classes `ZipInputStream` and `ZipOutputStream`, or am I wrong?

Beware that these classes are not part of the public API, so you would have to embed the full zipios code in your application. Otherwise you should use the `ZipFile` and `StreamEntry` classes as described above by Alexis.

**Comment by RokerHRO** [comment]
> > I though I could just use the classes `ZipInputStream` and `ZipOutputStream`, or am I wrong?
> 
> Beware that these classes are not part of the public API, so you would have to embed the full zipios code in your application. Otherwise you should use the `ZipFile` and `StreamEntry` classes as described above by Alexis.

They are not? How can I see that? They are declared in `/usr/include/zipios++/zipinputstream.h` and I consider all header files installed in this directory as part of the public API.

**Comment by mzf-guest** [comment]
As far as I understand, the public API are the files in the zipios folder in the source tree: https://github.com/Zipios/Zipios/tree/main/zipios and `zipinputstream.h` is not part of these files. Maybe @AlexisWilke would have more information?

**Comment by AlexisWilke** [comment]
Yes. Only the headers under zipios are considered public. Now anyone is welcome to use anything they want. It's open source software.

Also, the new library installs its headers under `/usr/include/zipios`. No "++" in the filename. We decided on that as the "++" can be misinterpreted by some compilers.

Under Ubuntu 22.04, 23.04, and 23.10, you can try installing with:

    sudo apt-get install libzipios-dev

You'll need to setup the PPA first. You should find the necessary instructions here:

https://launchpad.net/~snapcpp/+archive/ubuntu/ppa

Then you'll see the exact files accessible to you (public API, libraries, etc.)
